tokio/src/task/join_map.rs

```diff
@@ -0,0 +1,566 @@
+use crate::{
+    runtime::Handle,
+    task::{AbortHandle, JoinError, JoinHandle, LocalSet},
+    util::IdleNotifiedSet,
+};
+use std::{
+    borrow::Borrow,
+    collections::{
+        self,
+        hash_map::{HashMap, RandomState},
+    },
+    fmt,
+    future::Future,
+    hash::{BuildHasher, Hash},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// A collection of tasks spawned on a Tokio runtime, associated with hash map
+/// keys.
+///
+/// This type is very similar to the [`JoinSet`] type, with the addition of a
+/// set of keys associated with each task. These keys allow [cancelling a
+/// task][abort] or [multiple tasks][abort_matching] in the `JoinMap` based on
+/// their keys, or [test whether a task corresponding to a given key exists][contains] in the `JoinMap`.
+///
+/// In addition, when tasks in the `JoinMap` complete, they will return the
+/// associated key along with the value returned by the task, if any.
+///
+/// A `JoinMap` can be used to await the completion of some or all of the tasks
+/// in the map. The map is not ordered, and the tasks will be returned in the
+/// order they complete.
+///
+/// All of the tasks must have the same return type `V`.
+///
+/// When the `JoinMap` is dropped, all tasks in the `JoinMap` are immediately aborted.
+///
+/// **Note**: This is an [unstable API][unstable]. The public API of this type
+/// may break in 1.x releases. See [the documentation on unstable
+/// features][unstable] for details.
+///
+/// [`JoinSet`]: crate::task::JoinSet
+/// [unstable]: crate#unstable-features
+/// [abort]: fn@Self::abort
+/// [abort_matching]: fn@Self::abort_matching
+/// [contains]: fn@Self::contains_task
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rt", tokio_unstable))))]
+pub struct JoinMap<K, V, S = RandomState> {
+    /// A hash map of keys to `AbortHandle`s, used to cancel tasks by key.
+    ///
+    /// Note: this currently stores an `Option<AbortHandle>`, in order to
+    /// implement the `abort_matching` method without adding a temporary
+    /// allocation. Because we need to take the `AbortHandle`s by value to abort
+    /// them, we have to take them out of the map somehow. We cannot do that
+    /// while iterating without also removing the tasks that are _not_ being
+    /// aborted, which would then require that we store those K-V pairs someplace
+    /// and then put them back afterwards (requiring an allocation). This seems
+    /// not great.
```

Comment:
we certainly can; I just thought it was better for the `AbortHandle` API to take `self`, because --- unlike a `JoinHandle` --- there's nothing else an `AbortHandle` can do once the task has been aborted. it might be better to just change it to take `&self`, though.

Reply:
Taking `&self` is more convenient for the caller. E.g. it allows using it from the destructor of a struct without wrapping it in an `Option`.

Reply:
Yeah, I've changed it to do that. I think it's the right call.